lib/client-api/src/story_store.js

```diff
@@ -134,28 +134,48 @@ export default class StoryStore extends EventEmitter {
     delete _data[id];
   };
 
-  addStory({ id, kind, name: storyName, story, getDecorated, parameters = {} }) {
+  addStory(
+    { id, kind, name, storyFn: original, parameters = {} },
+    { getDecorators, applyDecorators }
+  ) {
     const { _data } = this;
 
     if (_data[id]) {
       logger.warn(stripIndents`
-      Story with id ${id} already exists in the store!
+        Story with id ${id} already exists in the store!
 
-      Perhaps you added the same story twice, or you have a name collision?
-      Story ids need to be unique -- ensure you aren't using the same names modolo url-sanitization.`);
+        Perhaps you added the same story twice, or you have a name collision?
+        Story ids need to be unique -- ensure you aren't using the same names modolo url-sanitization.
+      `);
     }
 
-    _data[id] = toChild({
+    const identification = {
+      id,
       kind,
-      name: storyName,
-      story,
+      name,
+      story: name, // legacy
+    };
+
+    // immutable original storyFn
+    const getOriginal = memoize(1)(() => original);
+
+    // lazily decorate the story when it's loaded
+    const getDecorated = memoize(1)(() => applyDecorators(getOriginal(), getDecorators()));
```

Comment:
`applyDecorators` just wraps the function right? Why not just run this right away and store the result?

Reply:
I guess the idea is be "lazy"? I'm not quite sure that composing functions is very expensive

Reply:
I founds that applyDecorators is getting kinda expensive when run on all stories on init.. It delays the rendering of the first meaningful paint on large storybooks.

You don't know how many decorators a user has, nor what they do.

Reply:
Ok, sure, seems reasonable